swap/src/network/request_response.rs

```diff
@@ -171,3 +184,92 @@ where
         Ok(())
     }
 }
+
+#[derive(Clone, Copy, Debug, Default)]
+pub struct OneShotCodec<P> {
+    phantom: PhantomData<P>,
+}
+
+#[async_trait]
+impl<P> RequestResponseCodec for OneShotCodec<P>
+where
+    P: Send + Sync + Clone + ProtocolName,
+{
+    type Protocol = P;
+    type Request = Request;
+    type Response = Response;
+
+    async fn read_request<T>(&mut self, _: &Self::Protocol, io: &mut T) -> io::Result<Self::Request>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        debug!("enter read_request");
+        let message = upgrade::read_one(io, BUF_SIZE)
+            .await
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
```

Comment:
The `ReadOneError` returned here actually [has an `Io` variant](https://docs.rs/libp2p/0.34.0/libp2p/core/upgrade/enum.ReadOneError.html) that you could just return. Only the `TooLarge` variant would need to be mapped and that one probably should be an `io::ErrorKind::Other`.